pkg/devicemapper/devmapper.go

```diff
@@ -750,51 +750,49 @@ func activateDevice(poolName string, name string, deviceID int, size uint64, ext
 	return nil
 }
 
-// CreateSnapDevice creates a snapshot based on the device identified by the baseName and baseDeviceId,
-func CreateSnapDevice(poolName string, deviceID int, baseName string, baseDeviceID int) error {
-	devinfo, _ := GetInfo(baseName)
-	doSuspend := devinfo != nil && devinfo.Exists != 0
-
-	if doSuspend {
-		if err := SuspendDevice(baseName); err != nil {
-			return err
-		}
-	}
-
+// CreateSnapDeviceRaw creates a snapshot device. Caller is assumed to resume and suspend the device.
+func CreateSnapDeviceRaw(poolName string, deviceID int, baseName string, baseDeviceID int) error {
 	task, err := TaskCreateNamed(deviceTargetMsg, poolName)
 	if task == nil {
-		if doSuspend {
-			ResumeDevice(baseName)
-		}
 		return err
 	}
 
 	if err := task.setSector(0); err != nil {
-		if doSuspend {
-			ResumeDevice(baseName)
-		}
 		return fmt.Errorf("devicemapper: Can't set sector %s", err)
 	}
 
 	if err := task.setMessage(fmt.Sprintf("create_snap %d %d", deviceID, baseDeviceID)); err != nil {
-		if doSuspend {
-			ResumeDevice(baseName)
-		}
 		return fmt.Errorf("devicemapper: Can't set message %s", err)
 	}
 
 	dmSawExist = false // reset before the task is run
 	if err := task.run(); err != nil {
-		if doSuspend {
-			ResumeDevice(baseName)
-		}
 		// Caller wants to know about ErrDeviceIDExists so that it can try with a different device id.
 		if dmSawExist {
 			return ErrDeviceIDExists
 		}
-
 		return fmt.Errorf("devicemapper: Error running deviceCreate (createSnapDevice) %s", err)
+	}
+
+	return nil
+}
+
+// CreateSnapDevice creates a snapshot based on the device identified by the baseName and baseDeviceId,
+func CreateSnapDevice(poolName string, deviceID int, baseName string, baseDeviceID int) error {
+	devinfo, _ := GetInfo(baseName)
+	doSuspend := devinfo != nil && devinfo.Exists != 0
 
+	if doSuspend {
+		if err := SuspendDevice(baseName); err != nil {
+			return err
+		}
```

Comment:
you wan't something like this:
```
func CreateSnapDevice(poolName string, deviceID int, baseName string, baseDeviceID int) error {
        devinfo, _ := GetInfo(baseName)
        doSuspend := devinfo != nil && devinfo.Exists != 0
        if doSuspend {
                if err := SuspendDevice(baseName); err != nil {
                        return err
                }
                defer ResumeDevice(baseName)
        }
       return CreateSnapDeviceRaw(poolName, deviceID, baseName, baseDeviceID)
}
```
IIRC the reason why we didn't do it like this was, the original code was not using `defer`.
https://github.com/docker/docker/blob/master/pkg/devicemapper/devmapper.go#L800-L804

and since this code is very tricky (it handles multiple conditions) we didn't wanted to change and introduce new errors. Also `defer` is very convenient at ignoring errors. if there is an error in `ResumeDevice(baseName)` it will ignore it, the current style of code still handles that.

Let me know if you feel strongly about it, and I will change it.